rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,252 @@
+// Copyright 2020 DCS Corporation, All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// DISTRIBUTION A. Approved for public release; distribution unlimited.
+// OPSEC #4584.
+//
+// Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
+// Part 252.227-7013 or 7014 (Feb 2014).
+//
+// This notice must appear in all copies of this file and its derivatives.
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcpputils/asserts.hpp"
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rcutils/filesystem.h"
+
+#include "rosbag2_cpp/logging.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/reindexer.hpp"
+
+#include "rosbag2_storage/storage_options.hpp"
+
+#ifdef WIN32
+// Import windows filesystem functionality
+#include <windows.h>
+#include <tchar.h>
+#include <stdio.h>
+#else
+// We're on a UNIX system. Import their filesystem stuff instead
+#include <dirent.h>
+#endif
+
+
+namespace rosbag2_cpp
+{
+Reindexer::Reindexer(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_(std::move(storage_factory)),
+  metadata_io_(std::move(metadata_io)) {}
+
+Reindexer::~Reindexer() {}
+
+bool Reindexer::compare_relative_file(
+  const rcpputils::fs::path & first_path, const rcpputils::fs::path & second_path)
+{
+  std::regex regex_rule(".*_(\\d+)\\.db3", std::regex_constants::ECMAScript);
+
+  std::smatch first_match;
+  std::smatch second_match;
+
+  auto first_path_string = first_path.string();
+  auto second_path_string = second_path.string();
+
+  auto first_regex_good = std::regex_match(first_path_string, first_match, regex_rule);
+  auto second_regex_good = std::regex_match(second_path_string, second_match, regex_rule);
+
+  // Make sure the paths have regex matches
+  if (!first_regex_good || !second_regex_good) {
+    throw std::runtime_error("Malformed relative file name. Expected numerical identifier.");
+  }
+
+  u_int32_t first_db_num = std::stoul(first_match.str(1), nullptr, 10);
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<rcpputils::fs::path> Reindexer::get_bag_files(
```

Comment:
nitpick: in C++ I generally like to avoid returning vectors by value - the API preference for me is

```
void get_bag_files(path base_folder, vector<path>& output) {}
```

In this case it probably doesn't matter, but that allows the caller to control resource allocation and reuse.